lasy/utils/mode_decomposition.py

```diff
@@ -0,0 +1,144 @@
+        in the decomposition. The keys of the dictionary are tuples
+        corresponding to (`n_x`,`n_y`)
+
+        waist : Beam waist for which the decomposition is calculated.
+            It is computed as the waist for which the weight of order 0 is maximum.
+    """
+
+    # Check if the provided laserProfile is a full laser profile or a
+    # transverse profile.
+
+    assert isinstance(
+        laserProfile, TransverseProfile
+    ), "laserProfile must be an instance of TransverseProfile"
+
+    # Get the field, sensible spatial bounds for the profile
+    lo = [None, None]
+    hi = [None, None]
+    if isinstance(laserProfile, TransverseProfileFromData):
+        lo[0] = laserProfile.field_interp.grid[0].min() + laserProfile.x_offset
+        lo[1] = laserProfile.field_interp.grid[1].min() + laserProfile.x_offset
+        hi[0] = laserProfile.field_interp.grid[0].max() + laserProfile.y_offset
+        hi[1] = laserProfile.field_interp.grid[1].max() + laserProfile.y_offset
+
+    else:
+        lo[0] = -laserProfile.w0 * 5 + laserProfile.x_offset
+        lo[1] = -laserProfile.w0 * 5 + laserProfile.x_offset
+        hi[0] = laserProfile.w0 * 5 + laserProfile.x_offset
+        hi[1] = laserProfile.w0 * 5 + laserProfile.x_offset
+
+    N_pts_x = int((hi[0] - lo[0]) / res)
+    N_pts_y = int((hi[1] - lo[1]) / res)
+
+    # Define spatial arrays
+    x = np.linspace(lo[0], hi[0], N_pts_x)
+    y = np.linspace(lo[1], hi[1], N_pts_y)
+    X, Y = np.meshgrid(x, y)
+    dx = x[2] - x[1]
+    dy = y[2] - y[1]
+
+    # Get the field on this grid
+    field = laserProfile.evaluate(X, Y)
+
+    # Get estimate of w0
+    w0 = estimate_best_HG_waist(x, y, field)
+
+    # Next we loop over the modes and calculate the relevant weights
+    weights = {}
+    for i in range(n_x_max):
+        for j in range(n_y_max):
+            HGMode = HermiteGaussianTransverseProfile(w0, i, j)
+            coef = np.sum(field * HGMode.evaluate(X, Y)) * dx * dy  # modalDecomposition
+            if math.isnan(coef):
+                coef = 0
+            weights[(i, j)] = coef
+
+    return weights, w0
+
+
+def estimate_best_HG_waist(x, y, field):
+    """
+    Estimate the waist that maximises the weighting of the first mode.
+
+    Calculates a D4Sigma waist as a first estimate and then tests multiple
+    gaussians with waists around this value to determine which has the best
+    overlap with the provided intensity profile. The aim here is to maximise
+    the energy in the fundamental mode of the reconstruction and so to avoid
+    a decomposition with significant higher-order modal content.
+
+    Parameters
+    ---------
+    x,y : 1D numpy arrays
+        representing the x and y axes on which the intensity profile is defined.
+
+    field : 2D numpy array representing the field (not the laser intensity).
+        the laser field profile in a 2D slice.
+
+    """
+
+    dx = x[2] - x[1]
+    dy = y[2] - y[1]
```

Comment:
Is this equivalent? Would help someone working with very small arrays...
```suggestion
    dx = x[1] - x[0]
    dy = y[1] - y[0]
```